Add Honeywell Lyric Component

timmo001 opened the pull request:
## Description:

Adds a new component to support Honeywell Lyric thermostat systems.

![image](https://user-images.githubusercontent.com/28114703/61334999-c9888580-a823-11e9-847c-b683f1b72aa0.png)

![image](https://user-images.githubusercontent.com/28114703/61334982-b83f7900-a823-11e9-9913-0a94a0afbf8e.png)

![image](https://user-images.githubusercontent.com/28114703/61335023-e0c77300-a823-11e9-9ebc-5469b2481a10.png)

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#9201

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] [_The manifest file_][manifest-docs] has all fields filled out correctly ([example][ex-manifest]).
  - [x] New dependencies have been added to `requirements` in the manifest ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

[ex-manifest]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/mobile_app/manifest.json
[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/mobile_app/manifest.json#L5
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard/__init__.py#L23
[manifest-docs]: https://developers.home-assistant.io/docs/en/development_checklist.html#_the-manifest-file_

timmo001 committed:
:sparkles: Add Honeywell Lyric Component

timmo001 committed:
:hammer: Remove hardcoded locations

timmo001 committed:
:hammer: Remove the lyric.lyric in the services

timmo001 committed:
:hammer: Add requirements from script

timmo001 committed:
:hammer: Added to coverage

awarecan commented:
Can we merged it to `honeywell` component? Especially since it is using `developer.honeywell.com` to generate an API key which is not limited to Lyric product line.

I have a Honeywell smart home security system integration on my todo list for a while which used the same developer account with Lyric.

timmo001 commented:
> Can we merged it to `honeywell` component? Especially since it is using `developer.honeywell.com` to generate an API key which is not limited to Lyric product line.
> 
> I have a Honeywell smart home security system integration on my todo list for a while which used the same developer account with Lyric.

Technically, these are completely different products. They may still have the Honeywell brand but they use totally different APIs. Merging this into the `honeywell` component would create a mess of three types of thermostat all mixed into one imo.

The Honeywell component also uses a totally different domain to the 'new' Honeywell T6 / T6R Lyric thermostats.

timmo001 committed:
:shirt: Fix linting errors

timmo001 committed:
:shirt: Fix lint error

timmo001 committed:
:shirt: Fix pylint warning

timmo001 committed:
:shirt: Fix pylinting

timmo001 committed:
:sparkles: Add next period time and status

timmo001 committed:
:hammer: Remove seconds in held status

bramkragten commented:
Thanks for picking up my work and making it ready for a PR.

timmo001 commented:
Ready for review, when someone gets the time to take a look 👍

devinslick commented:
Are there any changes pending or are we still waiting on a review?

balloob commented on homeassistant/components/lyric/__init__.py:
```diff
@@ -0,0 +1,186 @@
+}, extra=vol.ALLOW_EXTRA)
+
+
+def request_configuration(lyric, hass, config):
```
We shouldn't add new integrations using the configurator. Please use a config flow with an external step instead.

MartinHjelmare replied:
See the somfy integration for an example of a config flow like that.

timmo001 replied:
Thanks for the suggestion! Decided to rewrite the integration so it doesn't use the config and instead directly uses config flow. Much more easier to configure now just using the client id and secret then sending the user to honeywell to authenticate

balloob commented on homeassistant/components/lyric/__init__.py:
```diff
@@ -0,0 +1,186 @@
+
+DEPENDENCIES = ['http']
+
+REQUIREMENTS = ['python-lyric==1.1.4']
```
This should be moved to manifest.

balloob commented:
We are currently in a thermostat freeze until Climate 1.0 PR is merged. Besides waiting for that, this PR will also need to use config flows, config entries.

timmo001 committed:
:hammer: Remove requirements from init

timmo001 committed:
:hammer: :tada: Rewrite for config entry

timmo001 committed:
:hammer: Add missing unload

timmo001 committed:
:hammer: Cleanup

timmo001 committed:
:hammer: Remove name

timmo001 committed:
:hammer: Rework for device info

timmo001 committed:
:hammer: Model number

timmo001 committed:
:hammer: Cleanup

timmo001 committed:
:hammer: handled by frontend

timmo001 committed:
:ambulance: Fix climate integration

timmo001 commented:
Thanks for the pointers! 

Decided to rewrite the integration so it doesn't use the config and instead directly uses config flow. Much more easier to configure now just using the client id and secret then sending the user to Honeywell to authenticate. Much nicer process and can be configured with no effort. Really like this new flow system :1st_place_medal:

timmo001 committed:
:shirt: Cleanup and linting

timmo001 committed:
:books: Updated docs. Removed TODO

timmo001 committed:
:shirt: Add extra files to coveragerc

MartinHjelmare commented:
The climate remake is now done. Please update the PR to follow the new architecture.
https://developers.home-assistant.io/blog/2019/07/03/climate-cleanup.html

timmo001 committed:
:ambulance: Fix setup of token in flow

timmo001 committed:
:fire: Rewrite for climate 1.0

timmo001 commented:
OK, rewrote some more, updated to the new entity setup, asynced as much as I could see.

Works much nicer now. Like the new presets also. Allowed me to remove the `resume` service and use the preset to resume the current program instead, as well as permahold etc.

rblacklockjr commented:
Can you please assist me on how to add the lyric component using hassio on
docker?

On Tue., Jul. 16, 2019, 6:49 p.m. Timmo, <notifications@github.com> wrote:

> OK, rewrote some more, updated to the new entity setup, asynced as much as
> I could see.
>
> Works much nicer now. Like the new presets also. Allowed me to remove the
> resume service and use the preset to resume the current program instead,
> as well as permahold etc.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/home-assistant/pull/23015?email_source=notifications&email_token=ADYWLU6WM5KGGPDR53UFBCDP7ZGBLA5CNFSM4HFK6MAKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD2CPTIQ#issuecomment-512031138>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADYWLUYUR6NQTYTQSF2EETLP7ZGBLANCNFSM4HFK6MAA>
> .
>

timmo001 committed:
:hammer: Add extra preset

balloob commented on .coveragerc:
```diff
@@ -354,6 +354,11 @@ omit =
     homeassistant/components/lyft/sensor.py
+    homeassistant/components/lyric/__init__.py
+    homeassistant/components/lyric/climate.py
+    homeassistant/components/lyric/config_flow.py
```
Config flow needs tests

balloob commented on homeassistant/components/lyric/__init__.py:
```diff
@@ -0,0 +1,151 @@
+                else:
+                    _LOGGER.debug("Ignoring location %s, not in %s",
+                                  location.name, self._location)
+        except TypeError:
```
Type error is a very weird way to indicate a connection error

balloob commented on homeassistant/components/lyric/config_flow.py:
```diff
@@ -0,0 +1,152 @@
+            step_id='user',
+            data_schema=vol.Schema(
+                {
+                    vol.Required(CONF_CLIENT_ID): str,
```
Client ID and secret are not allowed to be asked via a config flow and need to be set via `configuration.yaml`. 

Example in Toon:
 - [config schema](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/toon/__init__.py#L43-L56)

timmo001 committed:
:hammer: Switch to connection error

timmo001 committed:
:hammer: Move client id and secret to domain config

timmo001 committed:
:hammer: Cleanup debugs

balloob commented on homeassistant/components/lyric/config_flow.py:
```diff
@@ -73,12 +61,19 @@ def __init__(self):
             self.hass.config.api.base_url, AUTH_CALLBACK_PATH)
         token_cache_file = self.hass.config.path(CONF_LYRIC_CONFIG_FILE)
 
+        _LOGGER.warning(client_id)
```
Debug ?

timmo001 replied:
Yup. Removed

timmo001 committed:
:hammer: Add some tests

timmo001 commented:
OK, I'm terrible at unit tests. Let me know if there is anything else I should add :+1:

timmo001 committed:
:hammer: Cleanup

timmo001 committed:
:hammer: Add missing test requirement

timmo001 committed:
:books: rtfm

MartinHjelmare commented on homeassistant/components/lyric/climate.py:
```diff
@@ -0,0 +1,301 @@
+        """Set operation mode."""
+        _LOGGER.debug('Set operation mode: %s', hvac_mode)
+
+        if hvac_mode in [HVAC_MODE_HEAT_COOL, HVAC_MODE_HEAT, HVAC_MODE_COOL,
```
`hvac_mode` is already validated by the service schema.

MartinHjelmare commented on homeassistant/components/lyric/climate.py:
```diff
@@ -0,0 +1,301 @@
+    def hvac_mode(self) -> str:
+        """Return hvac mode ie. heat, cool, off."""
+        hvac_mode = None
+        if self._mode in [HVAC_MODE_HEAT_COOL, HVAC_MODE_HEAT, HVAC_MODE_COOL,
```
It's fragile to couple the lyric api directly to the home assistant api. If the home assistant api changes, the integration would break until the lyric library is updated.

I recommend making two dicts that map between the home assistant modes and the lyric device modes.

MartinHjelmare commented on homeassistant/components/lyric/config_flow.py:
```diff
@@ -0,0 +1,142 @@
+        self.client_id = None
+        self.client_secret = None
+        self.code = None
+        pass
```
Please remove the `pass`.

MartinHjelmare commented on homeassistant/components/lyric/config_flow.py:
```diff
@@ -0,0 +1,142 @@
+    async def async_step_auth(self, code):
+        """Create an entry for auth."""
+        # Flow has been triggered from Lyric api
+        if isinstance(code, str):
```
Why do we check type? What other types would be allowed?

MartinHjelmare commented on homeassistant/components/lyric/config_flow.py:
```diff
@@ -0,0 +1,142 @@
+            url=url
+        )
+
+    async def _get_authorization_url(self):
```
Make this a callback instead since we don't need to await inside.

MartinHjelmare commented on homeassistant/components/lyric/config_flow.py:
```diff
@@ -0,0 +1,142 @@
+                      token_cache_file=token_cache_file)
+
+        # pylint: disable=pointless-statement
+        lyric.getauthorize_url
```
What happens here?

timmo001 replied:
I think this actually makes sure the lib sets up the api. The integration fails to work without it so I left it in. Not 100% sure why it is needed.

See: https://github.com/bramkragten/python-lyric/blob/master/lyric/__init__.py#L678

MartinHjelmare replied:
The private api instance attribute isn't initialized until accessing this property, which could be considered a bug.

It would good if the library would be improved to have a more natural setup flow, to avoid having to disable pylint warnings. Maybe ask about this in a library repo issue?

bramkragten replied:
You create an instance of Lyric twice, once in `_get_authorization_url` and then again in `async_step_creation` might want to store the instance of `_get_authorization_url` and you don't have to do this.

The upstream code is old and needs improvement so PR's are welcome :-)

timmo001 replied:
Thanks @bramkragten. Should look better now :+1:

MartinHjelmare commented on homeassistant/components/lyric/config_flow.py:
```diff
@@ -0,0 +1,142 @@
+
+        if 'code' not in request.query or 'state' not in request.query:
+            return web_response.Response(
+                text="Missing code or state parameter in " + request.url
```
Please use new style string formatting instead of string concatenation.

MartinHjelmare commented on homeassistant/components/lyric/sensor.py:
```diff
@@ -0,0 +1,116 @@
+                    self._state = 'Following Schedule'
+                elif status == 'HoldUntil':
+                    self._state = 'Held until {}'.format(
+                        self.device.nextPeriodTime[:-3])
```
What kind of time is this? Is it an absolute utc timestamp?

timmo001 replied:
Here is what this returns:

![image](https://user-images.githubusercontent.com/28114703/61577446-31b6c000-aadf-11e9-87e6-3e9e5e45cb76.png)

MartinHjelmare replied:
We want all times to be absolute utc. Please make sure.

MartinHjelmare commented on homeassistant/components/lyric/sensor.py:
```diff
@@ -0,0 +1,116 @@
+        if device.indoorTemperature:
+            devices.append(LyricSensor(
+                device, location, hass, 'indoorTemperature', 'Temperature',
+                hass.config.units.temperature_unit, 'mdi:thermometer'))
```
Please set the entity `device_class` property appropriately when possible. See available device classes here:
https://developers.home-assistant.io/docs/en/entity_sensor.html#available-device-classes

MartinHjelmare commented on homeassistant/components/lyric/sensor.py:
```diff
@@ -0,0 +1,116 @@
+from homeassistant.helpers.typing import HomeAssistantType
+from .const import DATA_LYRIC_CLIENT, DATA_LYRIC_DEVICES, DOMAIN
+
+DEPENDENCIES = ['lyric']
```
See comments for climate platform that applies for the sensor platform too.

timmo001 committed:
:books: Remove link

timmo001 committed:
:hammer: Suggested changes and cleanup

timmo001 committed:
:hammer: Cleanup

timmo001 committed:
:hammer: Tweak

MartinHjelmare commented on homeassistant/components/lyric/config_flow.py:
```diff
@@ -124,7 +118,8 @@ class LyricAuthCallbackView(HomeAssistantView):
         if 'code' not in request.query or 'state' not in request.query:
             return web_response.Response(
-                text="Missing code or state parameter in " + request.url
+                text="Missing code or state parameter in %s".format(
```
Use `{}` as format operator.

MartinHjelmare commented on homeassistant/components/lyric/sensor.py:
```diff
@@ -0,0 +1,130 @@
+                    if time <= datetime.now():
+                        date = date + timedelta(days=1)
+                        time = time.replace(day=date.day)
+                    state = time
```
When using the timestamp device class, the time should be absolute utc iso formatted.

timmo001 replied:
See updated code. This now returns:

![image](https://user-images.githubusercontent.com/28114703/61579581-e8766880-aafe-11e9-8944-a26bd43c20e3.png)

*(ui bug found btw)*

timmo001 committed:
:hammer: Merge and cleanup

timmo001 committed:
:hammer: Use util

timmo001 committed:
:hammer: Handle all in service

timmo001 committed:
:shirt: Linting

timmo001 commented on homeassistant/components/lyric/climate.py:
```diff
@@ -0,0 +1,301 @@
+        schema=HOLD_PERIOD_SCHEMA)
+
+
+async def async_unload_entry(
```
Can you elaborate?

What does this in __init__ do then?
https://github.com/home-assistant/home-assistant/pull/23015/files#diff-1685c84ca6f1d04786dc2db76731893bR75

MartinHjelmare replied:
That is the one we should use. The function here, will never be called.

Instead use the dispatch helper to signal a target function here in the platform. Send the signal from `async_setup_entry` in the component. Here in the platform connect the target function to the signal.

MartinHjelmare replied:
Technically we don't need to use the dispatch helper to remove services as the is reached via core (hass). You could just move the service name constant into const.py package module. Then you can import that where you need it and remove the service as needed.

timmo001 replied:
OK, I've moved the service unload into the same method in init to simplify things

timmo001 committed:
:hammer: Cleanup

timmo001 committed:
:hammer: Cleanup

timmo001 committed:
:hammer: Relative import

timmo001 committed:
:hammer: Move unload to init and service name to const

timmo001 committed:
:hammer: Fix naming

timmo001 committed:
:ambulance: Fix service remove

timmo001 committed:
:shirt: Linting

MartinHjelmare commented on homeassistant/components/lyric/__init__.py:
```diff
@@ -0,0 +1,177 @@
+                else:
+                    _LOGGER.debug("Ignoring location %s, not in %s",
+                                  location.name, self._location)
+        except ConnectionError:
```
Bump.

MartinHjelmare commented on homeassistant/components/lyric/climate.py:
```diff
@@ -0,0 +1,256 @@
+        elif entity_ids:
+            target_thermostats = [device for device in devices
+                                  if device.entity_id in entity_ids]
+        else:
```
This case can't happen, right?

MartinHjelmare commented on homeassistant/components/lyric/climate.py:
```diff
@@ -0,0 +1,256 @@
+
+        if entity_ids == 'all':
+            target_thermostats = devices
+        elif entity_ids:
```
Put the `else:` here instead.

MartinHjelmare commented on homeassistant/components/lyric/climate.py:
```diff
@@ -0,0 +1,256 @@
+            target_thermostats = devices
+
+        for thermostat in target_thermostats:
+            await thermostat.async_set_preset_period(time)
```
Should we use `asyncio.wait`?

MartinHjelmare commented on homeassistant/components/lyric/sensor.py:
```diff
@@ -0,0 +1,131 @@
+                    now = dt_util.utcnow()
+                    if time <= now.time():
+                        now = now + timedelta(days=1)
+                    state = dt_util.as_local(dt_util.as_utc(
```
The state should be utc iso formatted time.

timmo001 replied:
If I remove the `dt_util.as_local` the time is now UTC and the entity shows the correct time diff, however in the more info dialog shows this in UTC format instead of +1 which is where it should be where I am. Is this a bug in the frontend to report or am I formatting the state incorrectly?

![image](https://user-images.githubusercontent.com/28114703/63510596-dc534180-c4d6-11e9-8530-ff2a6775d78a.png)

![image](https://user-images.githubusercontent.com/28114703/63510577-cfcee900-c4d6-11e9-9de8-363a55efcb05.png)

MartinHjelmare commented on tests/components/lyric/test_config_flow.py:
```diff
@@ -0,0 +1,68 @@
+    flow.flow_id = '1234567890abcdef'
+
+    result = await flow.async_step_user()
+    assert result['type'] == data_entry_flow.RESULT_TYPE_EXTERNAL_STEP
```
We should test the whole flow until completion and all logic paths.

MartinHjelmare commented:
There's a handful of comments left to address and complete the tests of the whole config flow.

balloob commented:
@timmo001 ?

timmo001 commented:
Got some time off through to the bank holiday, so I'll go ahead and address these now

timmo001 committed:
:shirt: Format with black for new standards

timmo001 committed:
:hammer: Format as utc

timmo001 committed:
:fire: Fix entities in service

timmo001 committed:
:hammer: asyncio.wait

timmo001 committed:
:hammer: Add missing hold mode

balloob commented:
Make sure to run Black.

balloob commented on homeassistant/components/lyric/__init__.py:
```diff
@@ -0,0 +1,180 @@
+        """Init Lyric devices."""
+        self.lyric = lyric
+
+        if not lyric.locations:
```
It can happen now that `self._location` is never defined. Looks like you expect it to be an empty list in `dev devices`. I would remove this `if`. The list comprehension will then also result in empty list ?

balloob commented on homeassistant/components/lyric/__init__.py:
```diff
@@ -0,0 +1,180 @@
+            for location in self.lyric.locations:
+                if location.name in self._location:
+                    for device in location.thermostats:
+                        yield (location, device)
```
If it's doing I/O and you're making it a generator, you're getting in muddy waters. Because it's easy to pass a generator to async world without thinking about it, causing I/O to happen when async code iterates over the generator. Such bugs are difficult to catch.

Example where I did catch it https://github.com/home-assistant/home-assistant/blob/691e3f61415c1e74b93e7ceaa27d5ea4b7d06de3/homeassistant/helpers/entity_platform.py#L194-L201

balloob commented on homeassistant/components/lyric/climate.py:
```diff
@@ -0,0 +1,277 @@
+        self._unit = temp_unit
+
+        # Setup supported hvac modes
+        self._hvac_modes = [HVAC_MODE_OFF]
```
let's append `HVAC_MODE_OFF` at the end. Some UIs might use the sort mode that is in the state machine and it's weird if `off` is the first option instead of last.

balloob commented on homeassistant/components/lyric/climate.py:
```diff
@@ -0,0 +1,277 @@
+            self._hvac_modes.append(HVAC_MODE_COOL)
+
+        if device.can_heat and device.can_cool:
+            self._hvac_modes.append(HVAC_MODE_HEAT_COOL)
```
Will you need to add `SUPPORT_TEMP_RANGE` to supported features ?

balloob commented on homeassistant/components/lyric/climate.py:
```diff
@@ -0,0 +1,277 @@
+    async def async_set_hvac_mode(self, hvac_mode: str) -> None:
+        """Set hvac mode."""
+        _LOGGER.debug("Set hvac mode: %s", hvac_mode)
+        self.device.operationMode = LYRIC_HVAC_MODES[hvac_mode]
```
This looks like it is doing I/O but it's inside an async method that is not allowed. Please implement `def set_hvac_mode` instead. 

Also, if you control the library, please update it so that you need to call functions instead of setting properties. Setting properties shouldn't do I/O and is very easy to block our event loop.

balloob commented on homeassistant/components/lyric/climate.py:
```diff
@@ -0,0 +1,277 @@
+
+    async def async_set_preset_mode(self, preset_mode: str) -> None:
+        """Set preset (PermanentHold, HoldUntil, NoHold, VacationHold) mode."""
+        self.device.thermostatSetpointStatus = preset_mode
```
Same

balloob commented on homeassistant/components/lyric/climate.py:
```diff
@@ -0,0 +1,277 @@
+
+    async def async_set_preset_period(self, period: str) -> None:
+        """Set preset period (time)."""
+        self.device.thermostatSetpointHoldUntil(period)
```
Same

balloob commented on homeassistant/components/lyric/climate.py:
```diff
@@ -0,0 +1,277 @@
+        """Set preset period (time)."""
+        self.device.thermostatSetpointHoldUntil(period)
+
+    async def _lyric_update(self) -> None:
```
Is this updating the data too on property access?

balloob replied:
@timmo001 this comment is still open

balloob commented on homeassistant/components/lyric/config_flow.py:
```diff
@@ -0,0 +1,132 @@
+
+
+@config_entries.HANDLERS.register(DOMAIN)
+class LyricFlowHandler(ConfigFlow):
```
We can do this nowadays instead of using the decorator:

```suggestion
class LyricFlowHandler(ConfigFlow, domain=DOMAIN):
```

balloob commented on homeassistant/components/lyric/sensor.py:
```diff
@@ -0,0 +1,189 @@
+        """Get values from lyric."""
+        if self.device:
+            if self.key == "thermostatSetpointStatus":
+                status = getattr(self.device, self.key)
```
Is this doing I/O ?

timmo001 replied:
Not sure.. Don't think so..

timmo001 committed:
:hammer: Suggested changes

timmo001 committed:
:hammer: use decorator

Co-Authored-By: Paulus Schoutsen <paulus@home-assistant.io>

MartinHjelmare commented on homeassistant/components/lyric/climate.py:
```diff
@@ -0,0 +1,282 @@
+        _LOGGER.debug("target_thermostats: %s", target_thermostats)
+
+        for thermostat in target_thermostats:
+            await asyncio.wait({thermostat.async_set_preset_period(time)})
```
Since this call will end up on the executor thread pool, we should probably not use `asyncio.wait` now, to avoid flooding the pool with jobs.

Maybe good to add a code comment about this.

timmo001 replied:
Done

timmo001 committed:
:hammer: suggested change

timmo001 commented:
Why is the CI failing? I can run `scripts/check_format` with no issues on my side:

![image](https://user-images.githubusercontent.com/28114703/64908610-69667080-d6fa-11e9-88ef-42807f6c1b1d.png)

balloob commented on homeassistant/components/lyric/__init__.py:
```diff
@@ -0,0 +1,173 @@
+
+async def async_setup_entry(hass: HomeAssistantType, entry: ConfigEntry) -> bool:
+    """Set up Lyric from a config entry."""
+    conf = hass.data[DATA_LYRIC_CONFIG]
```
This can fail if `DOMAIN not in config and so `async_setup` returns early?

balloob replied:
this comment is still the case.

balloob commented on homeassistant/components/lyric/__init__.py:
```diff
@@ -0,0 +1,173 @@
+    def devices(self):
+        """Generate a list of thermostats and their location."""
+        for location in self.lyric.locations:
+            if location.name in self._location:
```
Not sure why there is this check if you assign all `location.name` on initialization ?

balloob commented on homeassistant/components/lyric/climate.py:
```diff
@@ -0,0 +1,282 @@
+    lyric = hass.data[DOMAIN][DATA_LYRIC_CLIENT]
+
+    try:
+        devices = lyric.devices()
```
if this is doing I/O, it should do an `await`. or you need to run it in the executor. You're inside an async context.

balloob replied:
```suggestion
        devices = await hass.async_add_executor_job(lyric.devices)
```

balloob commented on homeassistant/components/lyric/__init__.py:
```diff
@@ -0,0 +1,173 @@
+    def __init__(self, lyric):
+        """Init Lyric devices."""
+        self.lyric = lyric
+        self._location = [location.name for location in lyric.locations]
```
It looks like the `devices` method can raise `ConnectionError`. That means that `lyric.locations` probably does too? That means that initializing a `LyricClient` can only be done inside an async context.

balloob replied:
Generally you want to avoid doing I/O inside a constructor.

balloob commented on homeassistant/components/lyric/climate.py:
```diff
@@ -0,0 +1,282 @@
+        else:
+            temp = kwargs.get(ATTR_TEMPERATURE)
+        _LOGGER.debug("Set temperature: %s", temp)
+        self.device.temperatureSetpoint = temp
```
This is doing I/O and so should be done in the executor.

```suggestion
        await self.hass.async_add_executor_job(setattr, self.device, "temperatureSetpoint", temp)
```

balloob commented on homeassistant/components/lyric/config_flow.py:
```diff
@@ -0,0 +1,132 @@
+
+                self.hass.http.register_view(LyricAuthCallbackView())
+
+                url = self.lyric.getauthorize_url
```
This is doing I/O inside async context?

balloob commented on homeassistant/components/lyric/config_flow.py:
```diff
@@ -0,0 +1,132 @@
+
+    async def async_step_creation(self, user_input=None):
+        """Create Lyric api and entries."""
+        self.lyric.authorization_code(self.code, self.flow_id)
```
I/O inside async context

balloob commented on homeassistant/components/lyric/config_flow.py:
```diff
@@ -0,0 +1,132 @@
+    name = AUTH_CALLBACK_NAME
+    url = AUTH_CALLBACK_PATH
+
+    @staticmethod
```
Shouldn't be a static method.

balloob commented on homeassistant/components/lyric/sensor.py:
```diff
@@ -0,0 +1,189 @@
+    lyric = hass.data[DOMAIN][DATA_LYRIC_CLIENT]
+
+    try:
+        devices = lyric.devices()
```
I/O inside async context

balloob commented on homeassistant/components/lyric/sensor.py:
```diff
@@ -0,0 +1,189 @@
+    hass.data[DOMAIN][DATA_LYRIC_DEVICES] = devices
+
+    devices = []
+    for location, device in lyric.devices():
```
I/O inside async context

balloob commented on homeassistant/components/lyric/sensor.py:
```diff
@@ -0,0 +1,189 @@
+                )
+            )
+
+    async_add_entities(devices, True)
```
Passing `True` means that we update them all before adding them to HA. Is that what you want?

frenck commented:
Looks like the output is suppressed and only the return code matters for the `check_format` script (something we might be needing to look into).

Try running `black --fast homeassistant`

timmo001 committed:
:hammer: Remove useless check

timmo001 commented:
Tried that command on both the host and the devcontainer. Nothing changed.. :confused: 

```
root@90c27a441f6c:/workspaces/home-assistant# black --fast homeassistant
All done! ✨ 🍰 ✨
2614 files left unchanged.
```

timmo001 committed:
:hammer: Change to sync

Co-Authored-By: Paulus Schoutsen <paulus@home-assistant.io>

timmo001 committed:
:hammer: Make non-static

timmo001 committed:
:hammer: Move I/O to executor

Co-Authored-By: Paulus Schoutsen <paulus@home-assistant.io>

timmo001 committed:
:hammer: Dont update before adding entities

timmo001 committed:
:shirt: Format

timmo001 committed:
:shirt: Formatting

timmo001 commented:
> Tried that command on both the host and the devcontainer. Nothing changed.. 
> 
> ```
> root@90c27a441f6c:/workspaces/home-assistant# black --fast homeassistant
> All done! ✨ 🍰 ✨
> 2614 files left unchanged.
> ```

Turns out I'd missed the formatting on one of the tests :) Did a dodgy `black *` command to find it

timmo001 committed:
:ambulance: Fix call

pho3nixf1re commented:
What is the status of this now? Anything I can to do pitch in and help?

pho3nixf1re commented:
I attempted to test this out but the when I select the integration i get an `Aborted` modal instead of the configuration modal. I suppose that is unexpected. I copied the lyric component as-is to my `custom_components` folder. Is that enough to begin testing or do I need to tweak the files for them to work in `custom_components`?

balloob commented on homeassistant/components/lyric/config_flow.py:
```diff
@@ -0,0 +1,131 @@
+
+        conf = self.hass.data.get(DATA_LYRIC_CONFIG)
+
+        self.client_id = conf[CONF_CLIENT_ID]
```
Okay. You're probably want to shout at me when you read this, but I think I have a PR #27727 that will make it no longer needed to implement your own OAuth2 flow. Using this new helper, you will be able to get Home Assistant take care of account linking. You will just have to get the tokens passed to the Lyric lib afterwards once it is set up.

timmo001 replied:
Took a look at the somfy integration and replicated. Tested and it works perfectly:

![image](https://user-images.githubusercontent.com/28114703/67620205-4a302800-f7fc-11e9-9436-25e7ff694299.png)


It does save time and effort, just wish you'd had that helper there before I went to the initial effort :smile:

timmo001 committed:
:hammer: Rerun from rebase

timmo001 committed:
:ambulance: Force token expires to float

timmo001 committed:
:fire: Replace config_flow with helper

timmo001 commented on script/gen_requirements_all.py:
```diff
@@ -42,6 +42,151 @@
     "VL53L1X2",
 )
 
+TEST_REQUIREMENTS = (
```
Are these null and void now?

balloob replied:
Yeah.

timmo001 commented:
What do I have left to do to get this across the line?

There may be a few things arise from the amount of rebases (see the above comment) since lots has changed since this PR started.

timmo001 committed:
:fire: Fix

timmo001 commented:
OK something odd is now happening. So I can fetch the data as before, but any command returns a 500 error.. Could be lyric is having API issues or that something is now misconfigured.. :confused:

timmo001 committed:
:hammer: Add token

timmo001 committed:
:hammer: Remove bad test

timmo001 commented:
OK something odd is now happening. So I can fetch the data as before with and all the data is correct, but any command returns a 500 error.. Could be lyric is having API issues or that something is now misconfigured.. :confused:

LouisMT commented:
When trying out the latest version I get the following error:

```
Error setting up entry Lyric for lyric

Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/config_entries.py", line 192, in async_setup
    hass, self
  File "/config/custom_components/lyric/__init__.py", line 70, in async_setup_entry
    hass, entry
  File "/usr/src/homeassistant/homeassistant/helpers/config_entry_oauth2_flow.py", line 318, in async_get_config_entry_implementation
    implementation = implementations.get(config_entry.data["auth_implementation"])
KeyError: 'auth_implementation'
```

Do I have to change anything in my configuration to make the new OAuth helper work? Unfortunately I can't find the docs for the new helper yet!

balloob commented on homeassistant/components/lyric/__init__.py:
```diff
@@ -0,0 +1,188 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+SCAN_INTERVAL = timedelta(seconds=10)
```
This is really aggressive. We should not hit a cloud API that fast.

balloob commented on homeassistant/components/lyric/__init__.py:
```diff
@@ -0,0 +1,188 @@
+    )
+
+    implementation.token = entry.data[CONF_TOKEN]
+    implementation.token_cache_file = hass.config.path(CONF_LYRIC_CONFIG_FILE)
```
Why is there a token cache file and why don't we use `entry.data` to store things (which is the default for the local implementation)?

balloob replied:
Never mind this, let's get this merged and we can fix this in a future PR.

balloob commented on homeassistant/components/lyric/climate.py:
```diff
@@ -0,0 +1,284 @@
+    hass: HomeAssistantType, entry: ConfigEntry, async_add_entities
+) -> None:
+    """Set up Lyric thermostat based on a config entry."""
+    lyric = hass.data[DOMAIN][DATA_LYRIC_CLIENT]
```
could we scope this by `hass.data[DOMAIN][entry.entry_id]` so it could support multiple accounts in the future ?

balloob replied:
Never mind this, let's get this merged and we can fix this in a future PR.

balloob commented:
@timmo001 there are still some places where you are doing I/O inside async contexts. This needs to be addressed before we can merge this.

kvangent commented:
@timmo001 - Super excited to see this functionality. Your work is much appreciated. 

I had one small issue while running these as a custom component. One of my thermostats refused to load. I did some quick debugging and had to change `HEAT_COOL` to `AUTO` in 2 places:

- [climate.py L61](https://github.com/home-assistant/home-assistant/blob/02413ec469a5986b613243d7490e22220602080f/homeassistant/components/lyric/climate.py#L61) 
-  [climate.py L68](https://github.com/home-assistant/home-assistant/blob/02413ec469a5986b613243d7490e22220602080f/homeassistant/components/lyric/climate.py#L68) 

After that everything seems to work well. Thanks again 👍

pho3nixf1re commented:
Was this closed by mistake? Does it need someone to pick it up?

balloob commented:
It has gone stale, So yeah, someone needs to pick it up.

pho3nixf1re commented:
I really want this integration so I'll pick up the flag. If @timmo001 wants to continue to work on this feel free to let me know.

balloob commented:
That would be great ! You can fork his branch and open a new PR.

timmo001 commented:
Yes please someone do pick this up. Not had the time to keep track of this one. All working locally, just needs the above addressing which others will have more experience of implementing than myself :+1:

shaun-adam commented:
@pho3nixf1re did you end up creating a fork? I don't have a lot of coding depth but could help out as well.

pho3nixf1re commented:
I have not yet. I have been out of town for a bit over a week. I plan to dig in on this over the next few weekends around the holidays.

jasondefuria commented:
@pho3nixf1re Also wondering if you forked it. I was hoping I could help out as well. Same situation as @shaun-adam- not incredibly advanced at coding, but would love to help anyway I can!

degan6 commented:
I would love to see this merged. From what I can tell reading through the comments and other related pull requests this is ready to go but needed a fix to remove I/O operations in a async context? 

Also,#31387 was made after, but from what I've read on the requests this one seems to be better?

jasondefuria commented:
@degan6 Yes, also needs the get_url method too instead of deprecated base_url. I have been trying to play around and see if I can get this running.
